samza-core/src/main/java/org/apache/samza/operators/MessageStreamImpl.java

```diff
@@ -116,38 +124,49 @@ public K getKey(M message) {
       }
 
       @Override
-      public K getOtherKey(OM message) {
-        return joinFn.getSecondKey(message);
+      public KeyValueStore<K, PartialJoinMessage<M>> getState() {
+        return thisStreamState;
       }
 
       @Override
       public void init(Config config, TaskContext context) {
+        // joinFn#init() must only be called once, so we do it in this partial join function's #init.
         joinFn.init(config, context);
+
+        thisStreamState = new InternalInMemoryStore<>();
       }
     };
 
-    PartialJoinFunction<K, OM, M, RM> parJoin2 = new PartialJoinFunction<K, OM, M, RM>() {
+    PartialJoinFunction<K, JM, M, RM> otherPartialJoinFn = new PartialJoinFunction<K, JM, M, RM>() {
+      private KeyValueStore<K, PartialJoinMessage<JM>> otherStreamState;
+
       @Override
-      public RM apply(OM m1, M m) {
-        return joinFn.apply(m, m1);
+      public RM apply(JM om, M m) {
+        return joinFn.apply(m, om);
       }
 
       @Override
-      public K getKey(OM message) {
+      public K getKey(JM message) {
         return joinFn.getSecondKey(message);
       }
 
       @Override
-      public K getOtherKey(M message) {
```

Comment:
Nice! This method is no longer required. The newer API for `PartialJoinFn` is also more general.